ext/uri/uri_parser_whatwg.c

```diff
@@ -525,29 +513,49 @@ static zend_result php_uri_parser_whatwg_fragment_write(struct uri_internal_t *i
 
 PHP_RINIT_FUNCTION(uri_parser_whatwg)
 {
-	lexbor_mraw_t *mraw = lexbor_mraw_create();
-	lxb_status_t status = lexbor_mraw_init(mraw, lexbor_mraw_byte_size);
+	lxb_status_t status;
+	
+	status = lexbor_mraw_init(&lexbor_mraw, lexbor_mraw_byte_size);
 	if (status != LXB_STATUS_OK) {
-		lexbor_mraw_destroy(mraw, true);
-		return FAILURE;
+		goto fail;
 	}
 
-	status = lxb_url_parser_init(&lexbor_parser, mraw);
+	status = lxb_url_parser_init(&lexbor_parser, &lexbor_mraw);
 	if (status != LXB_STATUS_OK) {
-		lxb_url_parser_destroy(&lexbor_parser, false);
-		lexbor_mraw_destroy(mraw, true);
-		return FAILURE;
+		goto fail;
+	}
+
+	status = lxb_unicode_idna_init(&lexbor_idna);
+	if (status != LXB_STATUS_OK) {
+		goto fail;
 	}
 
 	parsed_urls = 0;
 
 	return SUCCESS;
+
+ fail:
+
+	/* Unconditionally calling the _destroy() functions is
+	 * safe on a zeroed structure. */
+	lxb_unicode_idna_destroy(&lexbor_idna, false);
+	memset(&lexbor_idna, 0, sizeof(lexbor_idna));
+	lxb_url_parser_destroy(&lexbor_parser, false);
+	memset(&lexbor_parser, 0, sizeof(lexbor_parser));
+	lexbor_mraw_destroy(&lexbor_mraw, false);
+	memset(&lexbor_mraw, 0, sizeof(lexbor_mraw));
+
+	return FAILURE;
 }
 
 PHP_RSHUTDOWN_FUNCTION(uri_parser_whatwg)
 {
-	lxb_url_parser_memory_destroy(&lexbor_parser);
+	lxb_unicode_idna_destroy(&lexbor_idna, false);
```

Comment:
I suppose the destroy logic may be separated to another function that gets called from RSHUTDOWN and the failure path in RINIT.
Anyway, this is only a small remark

Reply:
I considered that, but thought that being explicit here is useful, since the use-cases are a little different: The failure in RINIT might need to correctly handle partial initialization, whereas the cleanup in RSHUTDOWN can rely on proper initialization.